src/inmanta/protocol/openapi/converter.py

```diff
@@ -15,45 +15,72 @@
 
     Contact: code@inmanta.com
 """
+import inspect
 import json
-from typing import Dict, List, Type
+import uuid
+from types import FunctionType
+from typing import Dict, List, Optional, Union
 
 from pydantic.main import BaseModel
 
-from inmanta.protocol.common import ArgOption, UrlMethod
-from inmanta.protocol.openapi.model import OpenAPI, Operation, Parameter, ParameterType, PathItem, RequestBody, Response
+from inmanta import util
+from inmanta.protocol.common import ArgOption, MethodProperties, UrlMethod
+from inmanta.protocol.openapi.model import (
+    Header,
+    OpenAPI,
+    Operation,
+    Parameter,
+    ParameterType,
+    PathItem,
+    Reference,
+    RequestBody,
+    Response,
+)
 from inmanta.server import config
 from inmanta.util import get_compiler_version
 
 
-def openapi_json_encoder(o):
+def openapi_json_encoder(o) -> Union[Dict, str, List]:
     if isinstance(o, BaseModel):
         return o.dict(by_alias=True, exclude_none=True)
+    return util.custom_json_encoder(o)
 
 
 class OpenApiConverter:
     """
         Extracts API information for the OpenAPI definition from the server
     """
 
-    def get_servers(self):
+    def __init__(self, global_url_map: Dict[str, Dict[str, UrlMethod]]):
+        self.global_url_map = global_url_map
+        self.arg_option_handler = ArgOptionHandler()
+        self.type_converter = OpenApiTypeConverter()
+
+    def _get_servers(self) -> List[Dict[str, str]]:
         bind_port = config.get_bind_port()
         bind_addresses = config.server_bind_address.get()
         return [{"url": f"http://{bind_address}:{bind_port}/"} for bind_address in bind_addresses]
 
-    def get_top_level_object(self):
+    def _get_envelope(self) -> Dict[str, Dict]:
         info = {"title": "Inmanta Service Orchestrator", "version": get_compiler_version()}
-        servers = self.get_servers()
-        return {"openapi": "3.0.2", "info": info, "paths": {}, "servers": servers}
+        servers = self._get_servers()
+        paths = {}
+        for methods in self.global_url_map.values():
+            api_methods = self._filter_api_methods(methods)
+            if len(api_methods) > 0:
+                path = self._get_path_of_method(api_methods)
+                path_item = self._extract_operations_from_methods(api_methods, path)
+                paths[path] = path_item
+        return {"openapi": "3.0.2", "info": info, "paths": paths, "servers": servers}
 
-    def filter_api_methods(self, methods: Dict[str, UrlMethod]) -> Dict[str, UrlMethod]:
+    def _filter_api_methods(self, methods: Dict[str, UrlMethod]) -> Dict[str, UrlMethod]:
         return {
             method_name: url_method
             for method_name, url_method in methods.items()
             if "api" in url_method.properties.client_types
         }
 
-    def get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
+    def _get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
         path = [
```

Comment:
isn't this duplicate code wrt protocol? I expect it to be somewhere int there as well?